docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+We also need to consider the "clients" engine in this state - we will need to communicate commands sent by other devices and commands that we wish to send to other devices.
+
+We will rely on the consuming app to ensure that the sync manager is initialized correctly before syncing any engines.
+
+While we can "trust" each engine, we should try, as much as possible, to isolate engine data from other engines. For example, the keys used to encrypt a collection should only be exposed to that specific engine, and there's no need for one engine to know what info/collections returns for other engines.
+
+# Implementation plan
+
+* The sync manager is responsible for persisting the global syncID and version, considering the state to be "new" if it differs, and failing fatally if the version isn't acceptable. Engines never need to see these values.
+
+* The sync manager is responsible for persisting and populating the "declined" engine list. If an engine is declined, the manager should not ask it to sync. IOW, an engine doesn't really need to know whether it is enabled or not. However, engine should expose a way to be "reset" when the manager notices the declined state for an engine has been changed. This will allow engines to optimize certain things, such as not keeping tomstones etc. As part of this, the sync manager must expose an API so the consuming application can change the disabled engine state.
+
+* The sync manager is responsible for managing the "clients" collection - engines should never see details about that engine. However, engines will need the ability to be told about commands targetting that engine. The sync manager will also need to be able to accept commands to be delivered to other clients - although that needs more thought as the conditions in which a command need to be sent typically aren't managed by the engine itself - eg, when a "bookmark" restore is done.
+
+* Each engine is responsible for persisting engine-specific stuff which needs to be persisted. An obvious thing which needs to be persisted is the syncID and version string for the engine itself, so it can take the necessary action when these change.
+
+* The sync manager will manage keys and pass the relevant key into the engine. Keys for engines should not be exposed to engines other than the key it is for (although currently, in practice, all engines share the same key, but this is an implementation detail.
+
+* The sync manager will pass the relevant part of info/collections to the engine and a token which can be used to access storage. Specific engines never talk to the token server.
+
+# Details
+## Structure definitions
+
+While we use rust struct definitions here, it's important to keep in mind that as mentioned above, we'll need to support the manager being written in something other than rust. As the data is relatively small, it may (or may not) be desirable to use JSON as the serialization format.
```

Comment:
JSON vs not JSON is basically a question of: Do we have to write the glue in e.g. Kotlin/Swift to pull data out of JSON. If so, then probably using protobufs are easier and cleaner, since that glue can be generated automatically.

Reply:
I removed all refs to json.